### PR TITLE
Simplify and modernize build function implementations

### PR DESCRIPTION
**Pull request by Richienb**
I noticed that the same caching algorithm was implemented in all the build functions so I encapsulated it into its own function. Using promises also allowed the logic to be further simplified. I also have plans to simplify the other functions but creating updated branches of the upstream master is a bit of a hassle. Please add me to the org.

### CODE REVIEW

File: routes/spec.js [hunk 1]
```diff
@@ -1129,26 +1130,25 @@ Currently the API has a rate limit of **60 requests/minute** and **50,000 reques
...
-              }
-              return res.json(bazaar[itemId]);
-            });
-          });
+            if (itemId.includes(',')) {
+              return response.json(filterObject(bazaar, itemId.split(',')));
+            }
+            return response.json(bazaar[itemId]);
+          } catch (error) {
+            response.status(500).json({ error: error.message });
```
Comment (builder-247) [inline routes/spec.js:1150]: We probably want to just call the callback here and let the error handler middleware deal with it and not show the user the error.

**Comment by builder-247** [comment]
Can you do the same to the other endpoints as well :)

**Comment by Richienb** [comment]
@builder-247 Sure!

**Comment by Richienb** [comment]
@builder-247 All the build functions should now be complete. They just need a bit of testing.

**Comment by builder-247** [comment]
Getting a `buildBazaar is not a function` error

**Comment by Richienb** [comment]
@builder-247 What's the stack trace?

**Comment by Richienb** [comment]
Nevermind, found the issue.

**Comment by builder-247** [comment]
Looks like player stats object get sent as unhandled rejections

**Comment by Richienb** [comment]
@builder-247 Looks like I didn't include an `await`.

**Comment by builder-247** [comment]
pified mongoose functions don't get the correct context, getting ths error: 
```
MongooseError: `Model.findOneAndUpdate()` cannot run without a model as `this`. Make sure you are calling `MyModel.findOneAndUpdate()` where `MyModel` is a Mongoose model.
```

File: store/buildPlayer.js [hunk 2]
```diff
@@ -1,121 +1,75 @@
...
-  async.map(players, (player, done) => {
+async function populatePlayers(players) {
+  return async.map(players, async (player) => {
     const { uuid } = player;
-    queries.getPlayerProfile(uuid, (err, profile, isCached) => {
-      if (err) {
-        logger.error(err);
-      }
+    try {
+      const [profile, isCached] = pify(queries.getPlayerProfile, {
```
Comment (builder-247) [inline store/buildPlayer.js:51]: This seems to return a `pify is not a function` error

Comment (Richienb) [reply store/buildPlayer.js:51]: Is there a stack trace?

Comment (builder-247) [reply store/buildPlayer.js:51]:
```
at /usr/src/store/buildPlayer.js:51:35
     at /usr/src/node_modules/async/dist/async.js:148:38
     at /usr/src/node_modules/async/dist/async.js:243:13
     at eachOfArrayLike (/usr/src/node_modules/async/dist/async.js:500:13)
     at eachOf (/usr/src/node_modules/async/dist/async.js:551:16)
     at awaitable (/usr/src/node_modules/async/dist/async.js:208:32)
     at _asyncMap (/usr/src/node_modules/async/dist/async.js:241:16)
    at Object.map (/usr/src/node_modules/async/dist/async.js:594:16)
     at /usr/src/node_modules/async/dist/async.js:216:25
     at new Promise (<anonymous>)
     at Object.awaitable [as map] (/usr/src/node_modules/async/dist/async.js:211:20)
     at populatePlayers (/usr/src/store/buildPlayer.js:48:16)
|     at func (/usr/src/routes/spec.js:821:39)
     at processTicksAndRejections (internal/process/task_queues.js:97:5)

```

Comment (Richienb) [reply store/buildPlayer.js]: There was a missing `await` which caused the values to be extracted from the Promise object instead of the result.

**Comment by builder-247** [comment]
Error handling  behaviour got changed for players; instead of of getting an object with error key ad the appropriate status code, it now returns something like this:
```JSON
[
    {
        "status": 404,
        "message": "Invalid username!"
    }
]
```

**Comment by Richienb** [comment]
@builder-247 It looks like `async.map` puts the error object in an array. Destructuring should help.

**Comment by Richienb** [comment]
@builder-247 I fixed some functions which were incorrectly being asynchronous.

**Comment by Richienb** [comment]
@builder-247 It's better to be more explicit for non-primary function parameters.

File: routes/graphql.js [hunk 3]
```diff
@@ -107,15 +99,15 @@ class SkyblockResolver {
...
     const profiles = await redisGetAsync(`skyblock_profiles:${uuid}`);
     return profiles ? JSON.parse(profiles) : {};
   }
 
   async profile({ player_name, profile_id }) {
-    const uuid = await getUUIDAsync(player_name);
-    const profile = await buildProfileAsync(uuid, profile_id, true);
-    const players = await populatePlayersAsync(Object.keys(profile.members).map((uuid) => ({ uuid })));
+    const uuid = await getUUID(player_name);
+    const profile = await buildProfile(uuid, profile_id);
```
Comment (mattco98) [inline routes/graphql.js:109]: This call used to pass `true` for `shouldUpdateProfileList`, but doesn't anymore. Is that intentional?

File: routes/spec.js [hunk 4]
```diff
@@ -402,13 +407,13 @@ Currently the API has a rate limit of **60 requests/minute** and **50,000 reques
...
         },
         route: () => '/players/:player/quests',
-        func: (req, res) => {
-          getPlayer(req.params.player, (err, player) => {
-            if (err) {
-              return res.status(err.status).json({ error: err.message });
-            }
-            return res.json(player.quests);
-          });
+        func: async (req, res) => {
```
Comment (mattco98) [inline routes/spec.js:410]: Should be `request` and `response` to keep it consistent with the other changes

File: routes/spec.js [hunk 5]
```diff
@@ -618,13 +619,16 @@ Currently the API has a rate limit of **60 requests/minute** and **50,000 reques
           },
         },
         route: () => '/guilds/:player',
-        func: (req, res) => {
-          getGuildFromPlayer(req.params.player, req.query.populatePlayers, (err, guild) => {
-            if (err) {
-              return res.status(404).json({ error: err });
+        func: async (req, res, cb) => {
```
Comment (mattco98) [inline routes/spec.js:622]: Should be `request`, `response`, and `callback` to keep it consistent with the other changes

File: routes/spec.js [hunk 6]
```diff
@@ -808,27 +812,28 @@ Currently the API has a rate limit of **60 requests/minute** and **50,000 reques
...
-                  players.forEach((player) => {
-                    profile.members[player.profile.uuid].player = player.profile;
-                  });
-                  res.json(profile);
-                }
-              });
-            });
-          });
+            } catch (error) {
+              // TODO: Here also
```
Comment (mattco98) [inline routes/spec.js:831]: What is this `TODO` referring to?

File: routes/spec.js [hunk 7]
```diff
@@ -1128,27 +1133,26 @@ Currently the API has a rate limit of **60 requests/minute** and **50,000 reques
...
           },
         },
-        route: () => '/skyblock/bazaar/:id',
-        func: (req, res, cb) => {
-          const itemId = req.params.id;
-          redis.get('skyblock_bazaar', (err, resp) => {
-            const ids = JSON.parse(resp) || [];
-            if (itemId && !itemId.includes(',') && !ids.includes(itemId)) {
-              return res.status(400).json({ error: 'Invalid itemId' });
+        route: () => '/skyblock/bazaar/:id?',
```
Comment (mattco98) [inline routes/spec.js:1136]: Looks like a question mark sneaked in to the route here

File: package.json [hunk 8]
```diff
@@ -26,6 +26,7 @@
     "mongoose": "^5.9.10",
     "morgan": "^1.10.0",
     "nock": "^12.0.3",
+    "pify": "^5.0.0",
```
Comment (mattco98) [inline package.json:29]: Why is adding this dependency better than just using `util.promisify`?

File: routes/spec.js (hunk 7 above)
Comment (builder-247) [reply routes/spec.js:1136]: This is intentional, since the bazaar now returns all items if none are specified.

File: package.json (hunk 8 above)
Comment (Richienb) [reply package.json:29]: `pify` is faster and it supports multiple arguments returned by the callback.

File: routes/spec.js (hunk 6 above)
Comment (Richienb) [reply routes/spec.js]: That was a note to self fixed in https://github.com/slothpixel/core/pull/385/commits/96a7dc6444b8462bdc20a5d5980e0b49c55837ac.

File: routes/graphql.js (hunk 3 above)
Comment (Richienb) [reply routes/graphql.js:109]: That's because it's the default action.

File: routes/spec.js [hunk 9]
```diff
@@ -998,13 +998,13 @@ Currently the API has a rate limit of **60 requests/minute** and **50,000 reques
...
-        func: (req, res) => {
-          redis.get('skyblock_items', (err, items) => {
-            if (err) {
-              logger.error(err);
-              return res.status(500);
+        func: (_, response) => {
+          redis.get('skyblock_items', (error, items) => {
+            if (error) {
+              logger.error(error);
+              return response.status(500);
```
Comment (builder-247) [inline routes/spec.js:1005]: Let's use the callback here as well for consistency.

Comment (Richienb) [reply routes/spec.js]: Fixed!

File: routes/spec.js (hunk 4 above)
Comment (Richienb) [reply routes/spec.js]: Fixed!

File: routes/spec.js (hunk 5 above)
Comment (Richienb) [reply routes/spec.js]: Fixed!

**Pull request merged**